### PR TITLE
Use more folds to make more scala-like (avoiding allocations of collections when unnecessary)

### PR DESCRIPTION
**Pull request by EdgeCaseBerg**
Following up on Issue comments in #2 with some small refactors to avoid creation of collections that just get turned into `Array`s in order to use `qsum` among other things.

### CODE REVIEW

File: src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala [hunk 1]
```diff
@@ -17,17 +17,24 @@ case class OTFNAMETable(
...
-      Section("stringOffset", OTFOffset16(6 + tabeledRecords.map(_.length(b)).toArray.qsum.toInt)),
+      Section("stringOffset", OTFOffset16(6 + tabeledRecordsSum.toInt)),
       Section("nameRecords", OTFArray(tabeledRecords)),
       Section("data", strData)
     )
   }
 
-  override def length(b: ByteAllocator): UInt = UInt(6) + records.map(_(this).length(b)).toArray.qsum + strData.length(b)
+  override def length(b: ByteAllocator): UInt = {
+    records.foldLeft(UInt(6) + strData.length(b)) { // Start with 6 because ???
```
Comment (EdgeCaseBerg) [inline src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala]: Hey @TsundereBug what is the 6 here anyway?

Comment (TsundereBug) [reply src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala]: It's the length of the data in bytes before the record data.

Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala]: Remove comment here

File: src/main/scala/com/tsunderebug/scolor/otf/tables/OpenTypeTable.scala [hunk 2]
```diff
@@ -6,7 +6,12 @@ import spire.math.UByte
 
 abstract class OpenTypeTable extends Table {
 
-  override def getBytes(b: ByteAllocator): Array[UByte] = sections(b).flatMap(_.getBytes(b)).toArray
+  override def getBytes(b: ByteAllocator): Array[UByte] = {
+  	sections(b).foldLeft(Array.empty[UByte]) {
+  		case (accum, section) => accum ++ section.getBytes(b)
```
Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/tables/OpenTypeTable.scala]: 2 space indents are the code style for this project :)

File: src/main/scala/com/tsunderebug/scolor/otf/types/OTFArray.scala [hunk 3]
```diff
@@ -7,9 +7,17 @@ import spire.syntax.std.array._
 
 case class OTFArray[T <: Data](elems: Seq[T]) extends SectionDataType {
 
-  override def length(b: ByteAllocator): UInt = elems.map(_.length(b)).toArray.qsum
+  override def length(b: ByteAllocator): UInt = {
+  	elems.foldLeft(UInt(0)) {
+  		case (accum, elem) => accum + elem.length(b)
```
Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/types/OTFArray.scala]: 2 space indents

File: src/main/scala/com/tsunderebug/scolor/otf/types/OTFArray.scala [hunk 4]
```diff
@@ -7,9 +7,17 @@ import spire.syntax.std.array._
...
+  	}
+  }
 
-  override def getBytes(b: ByteAllocator): Array[UByte] = elems.flatMap(_.getBytes(b)).toArray
+  override def getBytes(b: ByteAllocator): Array[UByte] = {
+  	elems.foldLeft(Array.empty[UByte]) {
+  		case (accum, elem) => accum ++ elem.getBytes(b)
```
Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/types/OTFArray.scala]: 2 space indents

File: src/main/scala/com/tsunderebug/scolor/otf/tables/OTFCMAPTable.scala [hunk 5]
```diff
@@ -20,7 +20,9 @@ case class OTFCMAPTable(
   )
 
   override def length(b: ByteAllocator): UInt = {
-    UInt(sections(b).map(_.data.length(b).toLong).sum)
+    sections(b).foldLeft(UInt(0)) {
+      case (accum, section) => accum + section.data.length(b)
```
Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/tables/OTFCMAPTable.scala]: Bad use of foldLeft here

File: src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala [hunk 6]
```diff
@@ -17,17 +17,24 @@ case class OTFNAMETable(
 
   override def sections(b: ByteAllocator): Seq[Section] = {
     val off = b.allocate(this)
-    val dataOff = off.position + UInt(6) + tabeledRecords.map(_.length(b)).toArray.qsum
+    val tabeledRecordsSum = tabeledRecords.foldLeft(UInt(0)) {
+      case (accum, tabledRecord) => accum + tabledRecord.length(b)
```
Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala]: I like the tabeledRecordsSum, but again a bad use of foldleft

**Comment by TsundereBug** [comment]
What I mean by the above question is that does foldLeft save anything that sum/qsum/flatMap don't?

File: src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala [hunk 7]
```diff
@@ -17,17 +17,24 @@ case class OTFNAMETable(
...
+      Section("stringOffset", OTFOffset16(6 + tabeledRecordsSum.toInt)),
       Section("nameRecords", OTFArray(tabeledRecords)),
       Section("data", strData)
     )
   }
 
-  override def length(b: ByteAllocator): UInt = UInt(6) + records.map(_(this).length(b)).toArray.qsum + strData.length(b)
+  override def length(b: ByteAllocator): UInt = {
+    records.foldLeft(UInt(6) + strData.length(b)) { // 6 is the length of the data in bytes before the record data
+      case (accum, record) => accum + record(this).length(b)
```
Comment (TsundereBug) [inline src/main/scala/com/tsunderebug/scolor/otf/tables/OTFNAMETable.scala]: Again, what advantage is there of foldLeft over qsum? Memory?

**Comment by TsundereBug** [comment]
I read over the PR, and since it saves memory and CPU, seems like a good option :ok_hand:

**Comment by EdgeCaseBerg** [comment]
>avoiding allocations of collections when unnecessary

What I mean by that is in statements like this:

```
tabeledRecords.map(_.length(b)).toArray.qsum
```

You have a seq of `TableNameRecord`, then you have seq a collection of `UInt`'s after the map, and then you have an Array of UInts created from that Seq, and then finally you sum up the values in the array. So there's time spent copying data from one collection over into the other as the mapping and toArray-ing is done. (which can get slower/memory intensive with larger sequences)

But when you use foldLeft, you start out with the end type that you want, then do the mapping + length-getting all in one step. So it saves you a bit of memory usage. That's what I mean by avoiding allocation of collections if that makes more sense now? I don't know if the scala compiler has any tricks or optimizations it does under the hood to handle stuff like this since I imagine mapping + summing is a fairly common operation.

It makes me kinda wish there was a `sumBy` operation on lists defined in the general collection library. There is `view` though, which lazily applies map and other such functions, but I think that it doesn't get fully around the allocations I've mentioned because something like `toArray` will force evaluation of the view. Still in some cases I believe `view` would have the same benefits as `foldLeft` for example:

```
UInt(sections(b).map(_.data.length(b).toLong).sum)
```

when changed to 

```
UInt(sections(b).view.map(_.data.length(b).toLong).sum)
```

will avoid copying the collection from `sections(b)` and only actually apply the operation of getting the length of data + toLong while computing the sum, which is the same as foldLeft.

**Comment by TsundereBug** [comment]
Since `qsum` is a Spire thing, I might do a PR make it work on any GenTraversable. This way, it won't have to be evaluated with `toArray`.

**Pull request merged**